plutus-core/plutus-core/src/PlutusCore/Data.hs

```diff
@@ -78,27 +114,72 @@ toTerm = \case
     Constr i ds | otherwise         -> CBOR.TTagged 102 (CBOR.TList $ CBOR.TInteger i : fmap toTerm ds)
     Map es                          -> CBOR.TMap (fmap (bimap toTerm toTerm) es)
     List ds                         -> CBOR.TList $ fmap toTerm ds
-    I i                             -> CBOR.TInteger i
-    B b                             -> CBOR.TBytes b
+    I i                             -> integerToTerm i
+    B b                             -> bsToTerm b
+
+-- Logic for choosing encoding borrowed from Codec.CBOR.Write
+-- | Given an integer, create a 'CBOR.Term' that encodes it, following our size restrictions.
+integerToTerm :: Integer -> CBOR.Term
+-- If it fits in a Word64, then it's less than 64 bytes for sure, and we can just send it off
+-- as a normal integer for cborg to deal with
+integerToTerm i | i >= 0 , i <= fromIntegral (maxBound :: Word64) = CBOR.TInteger i
+                | i <  0 , i >= -1 - fromIntegral (maxBound :: Word64) = CBOR.TInteger i
+-- Otherwise, it would be encoded as a bignum anyway, so we manually do the bignum
+-- encoding with a bytestring inside, and since we use bsToTerm, that bytestring will
+-- get chunked up if it's too big.
+-- See Note [Evading the 64-byte limit]
+integerToTerm i | i >= 0 = CBOR.TTagged 2 $ bsToTerm (integerToBytes i)
+integerToTerm i | otherwise = CBOR.TTagged 3 $ bsToTerm (integerToBytes (-1 -i))
+
+-- Taken exactly from Codec.CBOR.Write
+integerToBytes :: Integer -> BS.ByteString
+integerToBytes n0
+  | n0 == 0   = BS.pack [0]
+  | otherwise = BS.pack (reverse (go n0))
+  where
+    go n | n == 0    = []
+         | otherwise = narrow n : go (n `shiftR` 8)
+
+    narrow :: Integer -> Word8
+    narrow = fromIntegral
+
+-- | Given an bytestring, create a 'CBOR.Term' that encodes it, following our size restrictions.
+bsToTerm :: BS.ByteString -> CBOR.Term
+bsToTerm b | BS.length b <= 64 = CBOR.TBytes b
+-- BSL.fromChunks preserves the chunks exactly as they are given (checked the source),
+-- and encoding a TBytesI encodes it with exactly the chunks in the lazy bytestring (checked the source),
+-- so this successfully results in our chunks being used on the wire.
+-- See Note [Evading the 64-byte limit]
+bsToTerm b                     = CBOR.TBytesI (BSL.fromChunks (to64ByteChunks b))
```

Comment:
Oh, wow. I see what you meant when you called this hacky. We're depending on a lot of implicit behavior here.

An alternative to consider:
If you use [`encodePreEncoded`](https://hackage.haskell.org/package/cborg-0.2.5.0/docs/Codec-CBOR-Encoding.html#v:encodePreEncoded), cborg will trust that whatever bytestring you pass in is valid cbor and emit that.

So something like
```
encodeBs b = encodePreEncoded (foldMap serialize (to64ByteChunks b))
```
would work.

This approach is unfortunately incompatible with the strategy of producing encodings via an intermediate `Term`, though. I think a lot would need to change to emit `Encoding` directly.

Reply:
Thanks, that's much better.

I think doing an `Encoding` directly is probably better at this point. At the beginning there was a very direct mapping to `Term`, which made it nice, that's now no longer really true.

Reply:
I think I managed to do it more simply: we can just encode each of the bytestrings in sequence, since the body of an indefinite-length bytestring just is a sequence of definite-length bytestrings.